rust/src/connection/network/transmitter/transaction.rs

```diff
@@ -241,60 +242,46 @@ impl TransactionTransmitter {
         mut on_close_callback_source: UnboundedReceiver<Box<dyn FnOnce(Option<Error>) + Send + Sync>>,
         mut shutdown_signal: UnboundedReceiver<()>,
     ) {
         const MAX_GRPC_MESSAGE_LEN: usize = 1_000_000;
-        const DISPATCH_INTERVAL: Duration = Duration::from_millis(3);
+        const DISPATCH_INTERVAL: Duration = Duration::from_micros(10);
 
         let mut request_buffer = TransactionRequestBuffer::default();
-        let mut next_dispatch = Instant::now() + DISPATCH_INTERVAL;
         loop {
-            select! { biased;
-                _ = shutdown_signal.recv() => {
-                    if !request_buffer.is_empty() {
-                        request_sink.send(request_buffer.take()).unwrap();
-                    }
-                    break;
+            if let Ok(_) = shutdown_signal.try_recv() {
+                if !request_buffer.is_empty() {
+                    request_sink.send(request_buffer.take()).unwrap();
                 }
-                _ = sleep_until(next_dispatch) => {
-                    if !request_buffer.is_empty() {
-                        request_sink.send(request_buffer.take()).unwrap();
-                    }
-                    next_dispatch = Instant::now() + DISPATCH_INTERVAL;
+                break;
+            } else if let Ok(recv) = request_source.try_recv() {
+                let (request, callback) = recv;
+                let request = request.into_proto();
+                if let Some(callback) = callback {
+                    collector.register(request.req_id.clone().into(), callback);
                 }
-                callback = on_close_callback_source.recv() => {
-                    if let Some(callback) = callback {
-                        collector.on_close.write().unwrap().push(callback)
-                    }
+                if request_buffer.len() + request.encoded_len() > MAX_GRPC_MESSAGE_LEN {
+                    request_sink.send(request_buffer.take()).unwrap();
                 }
-                recv = request_source.recv() => {
-                    if let Some((request, callback)) = recv {
-                        let request = request.into_proto();
-                        if let Some(callback) = callback {
-                            collector.register(request.req_id.clone().into(), callback);
-                        }
-                        if request_buffer.len() + request.encoded_len() > MAX_GRPC_MESSAGE_LEN {
-                            request_sink.send(request_buffer.take()).unwrap();
-                        }
-                        request_buffer.push(request);
-                    } else {
-                        break;
-                    }
+                request_buffer.push(request);
+            } else if let Ok(callback) = on_close_callback_source.try_recv() {
+                collector.on_close.write().unwrap().push(callback)
+            } else {
+                sleep(DISPATCH_INTERVAL);
```

Comment:
Wait, so we're not buffering here at all? Pseudocode:
```
loop {
    if shutdown { send(buffer); break; }
    else if request { buffer.push(request); }
    else if callback { register(callback); }
    else { sleep(10us); send(buffer); }
}
``` 
Unless the requests are sent down the channel before this thread gets the chance to sleep, we don't have more than one request at a time, and in either case the sleep is basically redundant. All it does is limit the send rate somewhat.

You could sleep for 1us, and send when dispatch interval is exceeded, but since there was apparently no downside to this I'd like to see the `async` version with no buffering explored as well. I'd prefer not spawning another thread if we can help it.

Reply:
Good point - I do want to keep the buffering because i know from way back when Haikal did his testing on throughput, without batching of messages we weren't able to maximise loading on the server side. What i'm changing it to is sleep for 1 microsecond, batching up to 20 microseconds worth of stuff (20 messages) , and then send.